src/snowflake/snowpark/_internal/analyzer/metadata_utils.py

```diff
@@ -22,31 +56,61 @@ def infer_metadata(
     from snowflake.snowpark._internal.analyzer.snowflake_plan import SnowflakePlan
     from snowflake.snowpark._internal.analyzer.unary_plan_node import (
         Filter,
+        Project,
         Sample,
         Sort,
     )
 
     attributes = None
+    quoted_identifiers = None
     # If source_plan is a LogicalPlan, SQL simplifier is not enabled
     # so we can try to infer the metadata from its child (SnowflakePlan)
     # When source_plan is Filter, Sort, Limit, Sample, metadata won't be changed
     # so we can use the metadata from its child directly
     if isinstance(source_plan, (Filter, Sort, Limit, Sample)):
         if isinstance(source_plan.child, SnowflakePlan):
             attributes = source_plan.child._attributes
+            quoted_identifiers = source_plan.child._quoted_identifiers
+    elif isinstance(source_plan, Project):
+        quoted_identifiers = infer_quoted_identifiers_from_expressions(
+            source_plan.project_list,  # type: ignore
+            analyzer,
+            df_aliased_col_name_to_real_col_name,
+        )
     # If source_plan is a SelectStatement, SQL simplifier is enabled
     elif isinstance(source_plan, SelectStatement):
         # When attributes is cached on source_plan, just use it
         if source_plan._attributes is not None:
             attributes = source_plan._attributes
+        # When _column_states.projection is available, just use it, which is
+        # 1) cached on self._snowflake_plan._quoted_identifiers
+        # 2) inferred in `derive_column_states_from_subquery` during `select()` call
+        if source_plan._column_states is not None:
+            quoted_identifiers = [c.name for c in source_plan._column_states.projection]
         # When source_plan.from_ is a Selectable and it doesn't have a projection,
         # it's a simple `SELECT * from ...`, which has the same metadata as it's child plan (source_plan.from_).
-        elif (
+        if (
             isinstance(source_plan.from_, Selectable)
             and source_plan.projection is None
             and source_plan.from_._snowflake_plan is not None
-            and source_plan.from_._snowflake_plan._attributes is not None
         ):
-            attributes = source_plan.from_.snowflake_plan._attributes
+            # only set attributes and quoted_identifiers if they are not set in previous step
+            if (
+                attributes is None
+                and source_plan.from_._snowflake_plan._attributes is not None
+            ):
+                attributes = source_plan.from_._snowflake_plan._attributes
+            if (
```

Comment:
i think we want to do elif here, because if the attributes is already set, we do not need to set up the quoted_identifier anymore

Reply:
yes good catch!